python/cuml/thirdparty_adapters/sparsefuncs_fast.py

```diff
@@ -0,0 +1,340 @@
+#
+# Copyright (c) 2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+import cupy as cp
+from numba import cuda
+from math import ceil
+
+
+def csr_mean_variance_axis0(X):
+    """Compute mean and variance
+    Parameters
+    ----------
+    X : sparse CSR matrix
+        Input array
+    Returns
+    -------
+    mean and variance
+    """
+    X = X.tocsc()
+    means, variances, _ = _csc_mean_variance_axis0(X)
+    return means, variances
+
+
+def csc_mean_variance_axis0(X):
+    """Compute mean and variance
+    Parameters
+    ----------
+    X : sparse CSC matrix
+        Input array
+    Returns
+    -------
+    mean and variance
+    """
+    means, variances, _ = _csc_mean_variance_axis0(X)
+    return means, variances
+
+
+def _csc_mean_variance_axis0(X):
+    """Compute mean, variance and nans count
+    Parameters
+    ----------
+    X : sparse CSC matrix
+        Input array
+    Returns
+    -------
+    mean, variance, nans count
+    """
+    n_samples, n_features = X.shape
+
+    means = cp.empty(n_features)
+    variances = cp.empty(n_features)
+    counts_nan = cp.empty(n_features)
+
+    start = X.indptr[0]
+    for i, end in enumerate(X.indptr[1:]):
```

Comment:
this is looping over columns, right? I wonder if in the future we should move this to CUDA. Is it in the critical path somewhere?

Reply:
Absolutely, this can be a real problem with large number of columns. I had the same problem for L1/L2 norms. I tried using CuPy streams without much success. Eventually, I instead opted for a conversion to Numba CUDA JIT code that was really efficient. To answer your question, yes we should probably move this to CUDA.